### DISCUSSION TITLE
Auxilio: problema ejercicio 2 actividad 8

### DISCUSSION THREAD

**Discussion by gaborszabo17** [Taller-Abierto-de-Humanidades-Digitales/curso_python_extenso#75] [Dudas del curso]
Hola, Jairo,
tengo este código:

`def fecha_nacimiento(katalog: dict, autor: str)->int:
    """ Retorna una lista con las fechas de nacimiento de los autores del catálogo.
     
     Parámetros
     ----------
     katalog: dict
        El catálogo en el que se desea buscar
     autor: str
        Nombre y apellido del autor del que se desea conocer su fecha de nacimiento
     
     Retorna
     ----------
     nacimiento: int
        La fecha de nacimiento del autor indicado"""
    nacimiento=0
    if len(autor.strip().split(" ")) < 2:
        return "Por favor, indique un nombre completo"
    else:
        for y in katalog.keys():
            if autor.strip().lower() in y.lower() or y.lower() in autor.strip().lower(): #Verificación de nombre
                    x=katalog[y]['bio']
                    if extraer_fechas(x) != []: #Verificación rutinaria
                        nacimiento= int(min(extraer_fechas(x))) #Esto asume que la mínima fecha en una biografía es la de nacimiento.   
                        return nacimiento 
                    else:
                        return f"No se extrajieron fechas de la biorafía de {autor}"   
            else:
                return (f"Parece que no se tiene biografías de {autor}")

print(fecha_nacimiento(biografias_boom, "mario llosa"))`

Cuando lo corro, lamentablemente me sale: 

`Parece que no se tiene biografías de mario llosa`

Lo que significa que el string "mario llosa" no cumple con el segundo if; pero no entiendo por qué no funciona.

¿Podrías ayudarme, por favor?

**Comment by jairomelo** [comment]
Hola @gaborszabo17 

El issue con tu código es que no estás aprovechando la lógica de los diccionarios :) 

Después del `else` de la primera condición, cuando pasas las claves realizas posteriormente dos pruebas más. El problema es que si falla en la primera evaluación, el loop se quiebra y no puede seguir. El algoritmo va más o menos así:

1. Evalúa si autor es menor a 2 palabras (True FIN | False pasa a 2)
2. Inicia loop con el listado de claves
3. **Evalúa si el argumento autor está en la clave 1 (True pasa a 5 | False pasa a 4)**
4. **Finaliza el loop y el programa**
5. Evalúa si Función extraer_fechas() no es una lista vacía (True pasa a 6 | False FIN)
6. Procesa listado de fechas, selecciona el valor menor, retorna valor
7. FIN

Es por eso que a menos que pongas "Julio Cortázar" en todos los demás casos va a fallar.

¿Cómo solucionas este problema? ¡Fácil! Usando la lógica del diccionario:

```
if len(autor.strip().split(" ")) < 2:
   return "Por favor, indique un nombre completo"
else:
    fechas = extraer_fechas(katalog[autor]['bio']) #<- buscas la clave del autor y en esta su bio, la pasas por `extraer_fechas()` y recibes el listado de fechas.
```

Incluso, si quieres asegurar que existan coincidencias del nombre del autor puedes aplicar el loop, pero no lo quiebres hasta encontrar una garantía de éxito o fracaso. POr ejemplo, puedes hacer algo como lo siguiente:

```
    if len(autor.strip().split(" ")) < 2:
        return "Por favor, indique un nombre completo"
    else:
        autor = [y for y in katalog.keys() if autor.strip().lower() in y.lower()][0] # <- Si hay una coincidencia parcial, regresa la clave correspondiente
        if autor: # Si el valor es nulo (es decir, no hay coincidencias) pasa al else
                bioautor = katalog[autor]["bio"]
                fechas = extraer_fechas(bioautor)
                return min(fechas)
        else:
                return f"No se encontró información sobre {autor}"
```

Espero que con estas indicaciones puedas resolver tu problema 😉
